### ISSUE TITLE
[Proposal] Quickstart Kotlin

### ISSUE THREAD

**Issue by rosariopfernandes** [firebase/quickstart-android#557]
It has been more than a year since [Kotlin](https://developer.android.com/kotlin/) became an official language for Android. It is being widely used by Android Developers and most of android samples and blog posts found on the internet are written in Kotlin nowadays.

I suggest rewriting, in Kotlin, the same android samples contained in this repository. Maybe pushing it to a new repository, since the java samples are also helpful.

**Comment by samtstern** [comment]
@rosariopfernandes thanks for bringing this up for discussion!

We are definitely working on bringing Kotlin to our Firebase developer materials, alongside Java.  The first thing we need to do is get our reference docs in order (since those are the most basic thing) and then we can move on to snippets and samples.

**Comment by the-dagger** [comment]
@samtstern Any tips on how folks can help the firebase team with the conversion?
Maybe starting off with the reference docs is something that we can help you out with?

**Comment by samtstern** [comment]
@the-dagger we're actually going to put reference docs on hold for now and jump straight into getting Kotlin snippets alongside the Java snippets in our guides.  I will be working with one of our tech writers int the very near future to start this.

I still have to think of a plan for getting Kotlin into this repo in a maintainable way that still makes it easy to navigate.  In `snippets-android` we defined common interface between Kotlin and Java classes but that doesn't work as well for an app.

**Comment by the-dagger** [comment]
I see, @samtstern that makes sense.
How doing something similar to what the official Android docs do?
They have separate packages for Android and Kotlin for the docs.

<img width="360" alt="screenshot 2018-07-24 at 10 06 03 pm" src="https://user-images.githubusercontent.com/6565270/43153098-fcf7d86c-8f8d-11e8-82d4-87111d98b5c3.png">

This will however be an issue with Dart arrives later on down the line, but happy to brainstorm more on this if you'd like to 👍

**Comment by samtstern** [comment]
@the-dagger yep that's the most likely course.  Can you link me to that repo?  Also what editor/theme are you using?  Very pretty haha.

**Comment by the-dagger** [comment]
Sure, here you go! 
https://github.com/android/snippets

Let me know if there's a way I can help, happy to do so 🙂 

For the theme, I'm using a chrome extension named Octotree, makes it super easy to browser github repos!
https://chrome.google.com/webstore/detail/octotree/bkhaagjahfmjljalopjnoealnfndnagc
<img width="357" alt="screenshot 2018-07-24 at 10 12 45 pm" src="https://user-images.githubusercontent.com/6565270/43153362-baa060d2-8f8e-11e8-8de2-d3afd6d17375.png">

**Comment by the-dagger** [comment]
@samtstern A gentle reminder on this one.
Really excited to see Kotlin support for the quickstart samples.

**Comment by samtstern** [comment]
Today I sat down with our writers and hashed out a plan to get Java and
Kotlin equality on all docs.  Once that's done (will take a bit of effort)
the quickstarts are next in line!

**Comment by the-dagger** [comment]
Awesome, that sounds a plan!
Looking forward to it, should be fun haha.

**Comment by samtstern** [comment]
@the-dagger @rosariopfernandes ok it's now time to start making this a reality.  But I think we need to think really carefully about how we want to do this.

Some options:

  * Create a whole new module for each quickstart in Kotlin.  So `auth` and `auth-kotlin`
  * Use a different build type / flavor for Kotlin
  * Build Kotlin and Java together and insert a first screen in the app where the user decides to run either
    the Kotlin or Java version
  * Other?

Some nice things I'd like to preserve:

  * Share non-code resources between Kotlin and Java quickstarts
  * Keep build times reasonable (local and on Travis)
  * Tools or interfaces or something to help us stay up to date and not let Java or Kotlin lag

**Comment by the-dagger** [comment]
@samtstern awesome, sounds great!
I would go with the following option : 

> Build Kotlin and Java together and insert a first screen in the app where the user decides to run either
> the Kotlin or Java version

This will not only be much easier to implement, but we can also use interfaces like we did with the snippets to ensure that the Kotlin/Java code snippets are always in sync with each other.
And as a bonus, we can also easily share resources among different activities.

@rosariopfernandes your thoughts?

**Comment by rosariopfernandes** [comment]
I think this

> Create a whole new module for each quickstart in Kotlin. So auth and auth-kotlin

sounds good, but I'm afraid people might run into issues like [this](https://github.com/firebase/snippets-android/issues/40) and not be able to fix it due to the large amount of modules in the project.

So I agree with @the-dagger . I think creating a first screen sounds like the best option.

**Comment by samtstern** [comment]
@rosariopfernandes @the-dagger you two are ridiculously productive.  Thank you for all the coding and reviewing today.  I'm going to leave for the weekend now, so I will pick this back up on Monday morning.

**SO** excited to see Kotlin happening here.  Without your help it would have taken a long time and the code quality would be lower.

**Comment by samtstern** [comment]
It's done!

@rosariopfernandes @the-dagger I can't say enough good things about your work here!  Not only is this a big win for Firebase developers, but I learned a ton about Kotlin from working with you.

**THANK YOU**

![ezgif-5-93ec62c484](https://user-images.githubusercontent.com/8466666/45560129-2167f180-b7f9-11e8-996b-f038a569098e.gif)

**Comment by the-dagger** [comment]
Always a pleasure, working on this was a lot of ```fun``` :D 

<img width="616" alt="screenshot 2018-09-14 at 9 12 56 pm" src="https://user-images.githubusercontent.com/6565270/45560519-f7e3b000-b862-11e8-9c25-0b16ae260a7a.png">

**Comment by rosariopfernandes** [comment]
Woo hoo! :tada: 
It's always a pleasure to help Firebase Developers and I've also learned a lot from you guys.
I just realized we've all been working on different timezones :sweat_smile: and it's amazing what we've accomplished in just a few days.

Now I can't wait to see the Kotlin Docs on the Firebase site!



_Dread it, Run from It, Kotlin still arrives_ :sparkles:

**Comment by samtstern** [comment]
Kotlin docs on firebase.google.com are (thanks to this work) no longer blocked by anything.  Unless something goes wrong we'll start adding Kotlin one product at a time over the next few weeks.

**Comment by samtstern** [comment]
@rosariopfernandes @the-dagger check it out!
https://firebase.google.com/docs/storage/android/start

All of the Cloud Storage Android docs now have Java / Kotlin side by side!  We're going to convert each product one-by-one, so more should be coming soon.

**Comment by rosariopfernandes** [comment]
Whoop Whoop! :confetti_ball:

**Comment by the-dagger** [comment]
This is great news!
Kudos to all of us. 🎉💯🚀

**Comment by samtstern** [comment]
Auth landed!
https://firebase.google.com/docs/auth/android/manage-users (and all other pages).

That was a big one, auth has a lot of snippets!

I am sure you'll stop caring at some point, but I will post here as each product gets Kotlin-ized for my own enjoyment haha

**Comment by samtstern** [comment]
FYI we're getting back to publishing more of the Kotlin docs (finally!).  Almost everything besides RTDB and Firestore are done and published, those two are coming soon as well.  

Example:
https://firebase.google.com/docs/ml-kit/android/use-custom-models

**Comment by the-dagger** [comment]
Yay!! 💯🎉

**Comment by samtstern** [comment]
Anddddddd it's done!  We got all the Kotlin snippets into all the pages.  The only stragglers might be the "setup" pages for each product since they use a strange format (and are consumed by Android Studio for tutorials) but even those will be mostly sorted out within a few days.

**Comment by rosariopfernandes** [comment]
Woo hoo! Awesome news. I'm glad our work is now public :laugh:
